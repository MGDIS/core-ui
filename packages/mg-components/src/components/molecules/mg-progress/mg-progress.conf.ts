import { AriaRole } from '../../../types';

/**
 * List of all possibles roles
 */
export const roles = ['progressbar', 'meter'] as const;

/**
 * Aria Role type from roles
 */
export type AriaRoleType = AriaRole & (typeof roles)[number];
