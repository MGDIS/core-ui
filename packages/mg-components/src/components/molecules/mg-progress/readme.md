# mg-progress



<!-- Auto Generated Below -->


## Properties

| Property             | Attribute    | Description                                                                                                                                                                                                     | Type                       | Default         |
| -------------------- | ------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------- | --------------- |
| `ariaRole`           | `aria-role`  | Define progress role - `progressbar` to indicate a progress, such as loading or percent completion of a task - `meter` to indicate a graphical display of a numeric value that varies within a defined range    | `"meter" \| "progressbar"` | `'progressbar'` |
| `label` _(required)_ | `label`      | Define label                                                                                                                                                                                                    | `string`                   | `undefined`     |
| `max`                | `max`        | Define the maximum value in the range..                                                                                                                                                                         | `number`                   | `100`           |
| `min`                | `min`        | Define the minimum value in the range..                                                                                                                                                                         | `number`                   | `0`             |
| `value`              | `value`      | Define current value By default, it will be displayed as a percentage value. If you don’t specify the min and max props, the value should be a number within the range of 0 to 100.                             | `number`                   | `0`             |
| `valueText`          | `value-text` | Define the value text override for assistive technologies. Assistive technologies read the value as a percentage by default. Use this prop to make the value more understandable and relevant. ex: "€20 of €50" | `string`                   | `undefined`     |


## CSS Custom Properties

| Name                               | Description                                                                                   |
| ---------------------------------- | --------------------------------------------------------------------------------------------- |
| `--mg-c-progress-border-radius`    | Defines the border radius of the progress bar. Default value is `--mg-b-size-32`.             |
| `--mg-c-progress-color`            | Defines the color of the progress bar. Default value is `--mg-b-color-app`.                   |
| `--mg-c-progress-color-background` | Defines the color of the progress bar background. Default value is `--mg-b-color-neutral-30`. |
| `--mg-c-progress-value`            | Defines the progress bar width. Default value is `0%`.                                        |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
