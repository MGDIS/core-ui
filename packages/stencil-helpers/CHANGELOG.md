# @mgdis/stencil-helpers

## 1.1.0

### Minor Changes

- a4091b3: Provide autocompletion

### Patch Changes

- fbdf99a: Update dependencies

## 1.0.4

### Patch Changes

- d6a9e03: Update dependencies

## 1.0.3

### Patch Changes

- 7a7e0ec: Update dependencies

## 1.0.2

### Patch Changes

- c87d711: Update dependencies

## 1.0.1

### Patch Changes

- eb3b1ab: Update dependencies

## 1.0.0

### Major Changes

- 50174ed: Stencil helpers package release

## 1.0.0

### Major Changes

- 50174ed: Stencil helpers package release
