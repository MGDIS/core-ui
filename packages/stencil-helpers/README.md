# @mgdis/stencil-helpers

This package centralizes helper functions for your [Stencil](https://stenciljs.com/) projects. It is divided into four thematic sections:

- **Components**: Centralizes all component utilities.
- **Locale**: Centralizes all necessary locale utilities for managing content translation.
- **Storybook**: Centralizes all Storybook helpers useful for building your own Storybook.
- **Test**: Centralizes all test utilities to simplify testing your components.

## Installation

```bash
pnpm i @mgdis/stencil-helpers
```

## Documentation

Read the full [documentation](./doc/README.md)
