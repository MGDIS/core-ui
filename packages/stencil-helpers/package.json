{
  "name": "@mgdis/stencil-helpers",
  "version": "3.2.1",
  "license": "ISC",
  "author": "Nicolas Jouanno <jouanno-n@mgdis.fr>",
  "maintainers": [
    "Simon Duhem <duhem-s@mgdis.fr>",
    "Nicolas Jouanno <jouanno-n@mgdis.fr>"
  ],
  "files": [
    "dist"
  ],
  "main": "./dist/index.umd.js",
  "module": "./dist/index.es.js",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.es.js",
      "require": "./dist/index.umd.js"
    }
  },
  "scripts": {
    "build": "vite build",
    "doc": "typedoc",
    "test": "TZ=UTC vitest",
    "coverage": "TZ=UTC vitest run --coverage"
  },
  "dependencies": {
    "@mgdis/core-ui-helpers": "workspace:*"
  },
  "devDependencies": {
    "@core-ui/tsconfig": "workspace:*",
    "@core-ui/typedoc-config": "workspace:*",
    "@core-ui/vitest-config": "workspace:*",
    "@types/node": "22.15.29",
    "@vitest/coverage-istanbul": "3.2.1",
    "@vitest/coverage-v8": "3.2.1",
    "htmlfy": "0.7.5",
    "typedoc": "0.28.5",
    "typedoc-plugin-markdown": "4.6.4",
    "vite": "6.3.5",
    "vite-plugin-dts": "4.5.4",
    "vitest": "3.2.1"
  }
}
