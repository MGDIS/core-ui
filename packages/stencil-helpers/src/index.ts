export * from './components';
export * from './ide';
export * from './locale';
export * from './storybook';
export * from './tests';
