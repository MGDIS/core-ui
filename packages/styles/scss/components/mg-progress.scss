/*!
 * @prop --mg-c-progress-border-radius: Defines the border radius of the progress bar. Default value is `--mg-b-size-32`.
 * @prop --mg-c-progress-color: Defines the color of the progress bar. Default value is `--mg-b-color-app`.
 * @prop --mg-c-progress-color-background: Defines the color of the progress bar background. Default value is `--mg-b-color-neutral-30`.
 * @prop --mg-c-progress-value: Defines the progress bar width. Default value is `0%`.
 */

:host {
  display: block;
}

.mg-c-progress {
  min-width: calc(var(--mg-b-size-16) * 4); // 64px

  &__bar {
    position: relative;
    overflow: hidden;
    height: var(--mg-b-size-8);
    box-sizing: content-box;
    border-width: var(--mg-b-size-4);
    border-style: solid;
    border-color: var(--mg-b-color-neutral-10);
    border-radius: var(--mg-c-progress-border-radius);
    background-color: var(--mg-c-progress-color-background);

    &::after {
      display: block;
      width: var(--mg-c-progress-value);
      height: 100%;
      border-radius: var(--mg-c-progress-border-radius);
      content: '';
      transition: width 1s;
    }

    &.mg-c-progress__bar--active::after {
      background-color: var(--mg-c-progress-bar-color-background);
      outline: var(--mg-b-color-neutral-10) solid var(--mg-b-size-4);
    }
  }
}
