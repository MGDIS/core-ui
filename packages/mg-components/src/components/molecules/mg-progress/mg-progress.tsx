import { Component, h, Host, Prop, Watch } from '@stencil/core';
import { isValidNumber, isValidString, toString } from '@mgdis/stencil-helpers';
import { AriaRoleType, roles } from './mg-progress.conf';

@Component({
  tag: 'mg-progress',
  styleUrl: '../../../../node_modules/@mgdis/styles/dist/components/mg-progress.css',
  shadow: true,
})
export class MgProgress {
  /**************
   * Decorators *
   **************/

  /**
   * Define label
   */
  @Prop() label!: string;
  @Watch('label')
  watchLabel(newValue: MgProgress['label']): void {
    if (!isValidString(newValue)) {
      throw new Error(`<mg-progress> prop "label" must be a valid string. Passed value: ${toString(newValue)}.`);
    }
  }

  /**
   * Define progress role
   * - `progressbar` to indicate a progress, such as loading or percent completion of a task
   * - `meter` to indicate a graphical display of a numeric value that varies within a defined range
   */
  @Prop() ariaRole?: AriaRoleType = 'progressbar';
  @Watch('ariaRole')
  watchAriaRole(newValue: MgProgress['ariaRole']): void {
    if (newValue && !roles.includes(newValue)) {
      throw new Error(`<mg-progress> prop "ariaRole" must be one of: ${roles.join(', ')}. Passed value: ${toString(newValue)}.`);
    }
  }

  /**
   * Define current value
   * By default, it will be displayed as a percentage value.
   * If you don’t specify the min and max props, the value should be a number within the range of 0 to 100.
   */
  @Prop() value?: number = 0;
  @Watch('value')
  watchValue(newValue: MgProgress['value']): void {
    if (newValue !== undefined && (!isValidNumber(newValue) || !(newValue >= this.min && newValue <= this.max)))
      throw new Error(`<mg-progress> prop “value” must be a number within the range ${this.min} to ${this.max}. Passed value: ${toString(newValue)}.`);
  }

  /**
   * Define the minimum value in the range
   */
  @Prop() min?: number = 0;
  @Watch('min')
  watchMin(newValue: MgProgress['min']): void {
    if (newValue !== 0 && newValue !== undefined && (!isValidNumber(newValue) || newValue >= this.max))
      throw new Error(`<mg-progress> prop “min” must be a number lower than: ${this.max}. Passed value: ${toString(newValue)}.`);
  }

  /**
   * Define the maximum value in the range
   */
  @Prop() max?: number = 100;
  @Watch('min')
  watchMax(newValue: MgProgress['max']): void {
    if (newValue !== 100 && newValue !== undefined && (!isValidNumber(newValue) || newValue <= this.min))
      throw new Error(`<mg-progress> prop “max” must be a number greater than: ${this.min}. Passed value: ${toString(newValue)}.`);
  }

  /*************
   * Lifecycle *
   *************/

  /**
   * Check if component props are well configured on init
   */
  componentWillLoad(): void {
    // validate props
    this.watchLabel(this.label);
    this.watchAriaRole(this.ariaRole);
    this.watchMin(this.min);
    this.watchMax(this.max);
    this.watchValue(this.value);
  }

  /**
   * Render
   * @returns HTML Element
   */
  render(): HTMLElement {
    return (
      <Host role={this.ariaRole} aria-label={this.label} aria-valuenow={this.value} aria-valuemin={this.min} aria-valuemax={this.max}>
        <div class="mg-c-progress">
          <div class={{ 'mg-c-progress__bar': true, 'mg-c-progress__bar--active': this.value > this.min }}>
            <div
              class="mg-c-progress__fill"
              style={{
                width: `${Math.round(((this.value - this.min) / (this.max - this.min)) * 100)}%`,
              }}
            ></div>
          </div>
        </div>
      </Host>
    );
  }
}
