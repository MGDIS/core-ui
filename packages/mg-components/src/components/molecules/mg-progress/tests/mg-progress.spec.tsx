import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgProgress } from '../mg-progress';
import { roles } from '../mg-progress.conf';
import { toString } from '@mgdis/stencil-helpers';

const getPage = (args: Pick<MgProgress, 'label'> & Partial<Pick<MgProgress, 'value' | 'min' | 'max' | 'valueText' | 'ariaRole'> & { lang: string }>) =>
  newSpecPage({
    components: [MgProgress],
    template: () => <mg-progress {...args}></mg-progress>,
  });

describe('mg-progress', () => {
  test.each([
    { label: 'progress' },
    { label: 'progress', value: 10 },
    { label: 'progress', value: 10, max: 10 },
    { label: 'progress', value: 10, max: 15 },
    { label: 'progress', value: 10, min: 5 },
    { label: 'progress', value: 10, min: 5, max: 15 },
    { label: 'progress', value: 30, max: 50, valueText: '30€ of 50€' },
    { label: 'progress', ariaRole: 'meter' as MgProgress['ariaRole'] },
    { label: 'new progress label' },
  ])('with args %s', async args => {
    const { root } = await getPage(args);
    expect(root).toMatchSnapshot();
  });

  test.each([' ', undefined, null])('Should throw error if "label" is not valid, %s', async label => {
    expect.assertions(1);
    try {
      await getPage({ label: label as MgProgress['label'] });
    } catch (err) {
      expect(err.message).toEqual(`<mg-progress> prop "label" must be a valid string. Passed value: ${label}.`);
    }
  });

  test.each(['batman', ' '])('Should throw error if "ariaRole" is not valid, %s', async ariaRole => {
    expect.assertions(1);
    try {
      await getPage({ label: 'progress', ariaRole: ariaRole as MgProgress['ariaRole'] });
    } catch (err) {
      expect(err.message).toEqual(`<mg-progress> prop "ariaRole" must be one of: ${roles.join(', ')}. Passed value: ${ariaRole}.`);
    }
  });

  test.each(['batman', 200, {}])('Should throw error if "value" is not valid, %s', async value => {
    expect.assertions(1);
    try {
      await getPage({ label: 'progress', value: value as MgProgress['value'] });
    } catch (err) {
      expect(err.message).toEqual(`<mg-progress> prop “value” must be a number within the range 0 to 100. Passed value: ${typeof value === 'string' ? 'NaN' : toString(value)}.`);
    }
  });

  test.each(['batman', 200, {}])('Should throw error if "min" is not valid, %s', async min => {
    expect.assertions(1);
    try {
      await getPage({ label: 'progress', min: min as MgProgress['min'] });
    } catch (err) {
      expect(err.message).toEqual(`<mg-progress> prop “min” must be a number lower than: 100. Passed value: ${typeof min === 'string' ? 'NaN' : toString(min)}.`);
    }
  });

  test.each(['batman', -10, {}])('Should throw error if "max" is not valid, %s', async max => {
    expect.assertions(1);
    try {
      await getPage({ label: 'progress', max: max as MgProgress['max'] });
    } catch (err) {
      expect(err.message).toEqual(`<mg-progress> prop “max” must be a number greater than: 0. Passed value: ${typeof max === 'string' ? 'NaN' : toString(max)}.`);
    }
  });

  test('Should throw error if valueText is not valid', async () => {
    expect.assertions(1);
    try {
      await getPage({ label: 'label', valueText: ' ' });
    } catch (err) {
      expect(err.message).toEqual(`<mg-progress> prop "valueText" must be a valid string. Passed value:  .`);
    }
  });
});
