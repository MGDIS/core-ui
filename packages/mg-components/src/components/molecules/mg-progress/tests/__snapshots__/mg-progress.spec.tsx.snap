// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-progress with args { label: 'new progress label' } 1`] = `
<mg-progress aria-busy="true" aria-label="new progress label" aria-valuemax="100" aria-valuemin="0" aria-valuenow="0" role="progressbar">
  <template shadowrootmode="open">
    <div class="mg-c-progress">
      <div class="mg-c-progress__bar" style="--mg-c-progress-value: 0%;"></div>
    </div>
  </template>
</mg-progress>
`;

exports[`mg-progress with args { label: 'progress' } 1`] = `
<mg-progress aria-busy="true" aria-label="progress" aria-valuemax="100" aria-valuemin="0" aria-valuenow="0" role="progressbar">
  <template shadowrootmode="open">
    <div class="mg-c-progress">
      <div class="mg-c-progress__bar" style="--mg-c-progress-value: 0%;"></div>
    </div>
  </template>
</mg-progress>
`;

exports[`mg-progress with args { label: 'progress', ariaRole: 'meter' } 1`] = `
<mg-progress aria-label="progress" aria-valuemax="100" aria-valuemin="0" aria-valuenow="0" role="meter">
  <template shadowrootmode="open">
    <div class="mg-c-progress">
      <div class="mg-c-progress__bar" style="--mg-c-progress-value: 0%;"></div>
    </div>
  </template>
</mg-progress>
`;

exports[`mg-progress with args { label: 'progress', value: 10 } 1`] = `
<mg-progress aria-busy="true" aria-label="progress" aria-valuemax="100" aria-valuemin="0" aria-valuenow="10" role="progressbar">
  <template shadowrootmode="open">
    <div class="mg-c-progress">
      <div class="mg-c-progress__bar mg-c-progress__bar--active" style="--mg-c-progress-value: 10%;"></div>
    </div>
  </template>
</mg-progress>
`;

exports[`mg-progress with args { label: 'progress', value: 10, max: 10 } 1`] = `
<mg-progress aria-busy="false" aria-label="progress" aria-valuemax="10" aria-valuemin="0" aria-valuenow="10" role="progressbar">
  <template shadowrootmode="open">
    <div class="mg-c-progress">
      <div class="mg-c-progress__bar mg-c-progress__bar--active" style="--mg-c-progress-value: 100%;"></div>
    </div>
  </template>
</mg-progress>
`;

exports[`mg-progress with args { label: 'progress', value: 10, max: 15 } 1`] = `
<mg-progress aria-busy="true" aria-label="progress" aria-valuemax="15" aria-valuemin="0" aria-valuenow="10" role="progressbar">
  <template shadowrootmode="open">
    <div class="mg-c-progress">
      <div class="mg-c-progress__bar mg-c-progress__bar--active" style="--mg-c-progress-value: 67%;"></div>
    </div>
  </template>
</mg-progress>
`;

exports[`mg-progress with args { label: 'progress', value: 10, min: 5 } 1`] = `
<mg-progress aria-busy="true" aria-label="progress" aria-valuemax="100" aria-valuemin="5" aria-valuenow="10" role="progressbar">
  <template shadowrootmode="open">
    <div class="mg-c-progress">
      <div class="mg-c-progress__bar mg-c-progress__bar--active" style="--mg-c-progress-value: 5%;"></div>
    </div>
  </template>
</mg-progress>
`;

exports[`mg-progress with args { label: 'progress', value: 10, min: 5, max: 15 } 1`] = `
<mg-progress aria-busy="true" aria-label="progress" aria-valuemax="15" aria-valuemin="5" aria-valuenow="10" role="progressbar">
  <template shadowrootmode="open">
    <div class="mg-c-progress">
      <div class="mg-c-progress__bar mg-c-progress__bar--active" style="--mg-c-progress-value: 50%;"></div>
    </div>
  </template>
</mg-progress>
`;

exports[`mg-progress with args { label: 'progress', value: 30, max: 50, valueText: '30€ of 50€' } 1`] = `
<mg-progress aria-busy="true" aria-label="progress" aria-valuemax="50" aria-valuemin="0" aria-valuenow="30" aria-valuetext="30€ of 50€" role="progressbar">
  <template shadowrootmode="open">
    <div class="mg-c-progress">
      <div class="mg-c-progress__bar mg-c-progress__bar--active" style="--mg-c-progress-value: 60%;"></div>
    </div>
  </template>
</mg-progress>
`;
